WIP: MDS Spec versus MDS Implementation

ericdill opened the pull request:
- based on conversations with @danielballan, @tacaswell, @arkilic 
  What other changes do we need guys?

ericdill committed:
WIP: Updates to mds spec

- based on conversations with @danielballan, @tacaswell, @arkilic

ericdill commented:
iirc the `event_descriptor` does not have a `type` key anymore, right?

ericdill commented:
oh, meant to loop @dchabot in on this too, sorry!

ericdill committed:
MNT: Minor changes to spec based on discussions

stuwilkins commented:
Hi Eric,

If you look at the top of the document it explains why types are required. While the trigger has gone (I am still a little uncomfortable about that, but @dchabot convinced me) One could think about putting this in the top run header, but I believe this reduces flexibility.

I would suggest not making changes without the small group getting together. @tacaswell made some really good points about how the current metadatastore can be used for different “stages” of data and building in that flexibility at this stage is really important.

Please call a meeting of the few if you would like to change things.

Thanks,
Stuart

On Jan 28, 2015, at 5:38 PM, Eric Dill <notifications@github.com<mailto:notifications@github.com>> wrote:

iirc the event_descriptor does not have a type key anymore, right?

—
Reply to this email directly or view it on GitHubhttps://github.com/NSLS-II/docs/pull/2#issuecomment-71931893.

ghost commented:
It was never part of the descriptor.

On Jan 28, 2015, at 6:03 PM, dchabot <notifications@github.com<mailto:notifications@github.com>> wrote:

On Wed, Jan 28, 2015 at 5:38 PM, Eric Dill <notifications@github.com<mailto:notifications@github.com>> wrote:

> oh, meant to loop @dchabot https://github.com/dchabot in on this too,
> sorry!
> 
> I've been following.
> But, I do not recall when the 'type' key was scheduled for execution.
> Was it ever included to begin with? Is it of any use 'downstream' (i.e.
> clients)?

—
Reply to this email directly or view it on GitHubhttps://github.com/NSLS-II/docs/pull/2#issuecomment-71935560.

tacaswell commented on source/arch/metadatastore-format.rst:
```diff
@@ -171,9 +170,9 @@ above ``event`` is described by the ``event_descriptor`` ::
             "chan7" : {"source" : "PV:XF:23ID1-ES{Sclr:1}.S7"},
             "chan8" : {"source" : "PV:XF:23ID1-ES{Sclr:1}.S8"}, 
-            "pimte: : {"source" : "FILESTORE:<...>"}
+            "pimte: : {"source" : "name_of_detector", "external": "FILESTORE:<...>"}
```
I thought this is what we decided talking with @stuwilkins

arkilic replied:
I think this document is definitely not concrete and always up for discussion. I wouldn't assume anything here is remotely solidified.

tacaswell commented on source/arch/metadatastore-format.rst:
```diff
@@ -161,7 +161,6 @@ above ``event`` is described by the ``event_descriptor`` ::
 
     {
         "uid" : <unique_id>,
-        "type" : "measure",
```
I am also not convinced that 'type' makes sense as the name.  The model provides information about what kind of event it is in other ways (what db/collection it is for one thing), type shadows all the things, and does not provide a clear ordering.

I think we should adopt the NASA language here and call this 'data_level' or some such where 0 is raw data.

tacaswell commented on source/arch/metadatastore-format.rst:
```diff
@@ -82,7 +82,7 @@ event. For example::
     {
         "uid" : <unique_id>,
-        "scan_id" : <non-unique-id>,
+        "scan_id" : <non-unique-id>, # is this specified to be an integer, or is anything valid here?
```
I would go with anything sortable

stuwilkins replied:
+1

On Jan 29, 2015, at 11:10 AM, Thomas A Caswell <notifications@github.com<mailto:notifications@github.com>> wrote:

In source/arch/metadatastore-format.rsthttps://github.com/NSLS-II/docs/pull/2#discussion_r23778463:

> @@ -82,7 +82,7 @@ event. For example::
> 
> ```
>  {
>      "uid" : <unique_id>,
> ```
> -        "scan_id" : <non-unique-id>,
> -        "scan_id" : <non-unique-id>, # is this specified to be an integer, or is anything valid here?

I would go with anything sortable

—
Reply to this email directly or view it on GitHubhttps://github.com/NSLS-II/docs/pull/2/files#r23778463.

ericdill committed:
ENH: Fleshed out description of event_desc keys

- Improved the description of the event descriptor keys
  dictionary. Done with @tacaswell

danielballan commented:
I am sitting beside Arman as he finishes implementing this. Here are some decisions we made that deviate from the documented spec:
- Event descriptors are documented to have `keys`, but this clobbers the Python method, so we are reverting to `data_keys`. We don't remember any discussion on this point anyway.
- Sample is currently implemented as a simple dict, not a full-blown collection like Beamline Config. Perhaps in the medium-term it should be an `DynamicEmbeddedDocument`.
- Begin Run Events are indexed on `_id`, `uid`, `time`, `owner`, and `scan_id`. End Run Events are indexed on `_id`, `time`, `exit_status`, and `begin_run_event`.
- Arman prefers the name `exit_status` to `completion_state` and I agree. Can revert later if others dissent.

tacaswell commented:
These specs shouldn't say anything about indexing

tacaswell commented:
also, work against #3 not this one
